### ISSUE TITLE
Subscribing other ETW events

### ISSUE THREAD

**Issue by yusufozturk** [rabbitstack/fibratus#99]
Hi @rabbitstack 

Is it possible to subscribe other ETW events via Fibratus?

![image](https://user-images.githubusercontent.com/928600/144727828-05f151ac-9cb0-4140-997a-6021dce30ac0.png)

For example, I would like to subscribe for DNS client events which is {1C95126E-7EEA-49A9-A3FE-A378B03DDB4D}. Does Fibratus support that today or needs to be implemented to support other event types?

Thanks.

**Comment by rabbitstack** [comment]
Hi @yusufozturk 

It doesn't support out of the box, but it could be easily extended for such use cases.
I was looking at the DNS client ETW provider, but my intention was to preserve the semantics of the network events and augment them with DNS or even HTTP payloads. So, for example, the `Send` event would contain among other parameters, the DNS query information.

**Comment by yusufozturk** [comment]
Hi Nedim,

Is it possible to provide one example to subscribe an ETW event where we can get the details dynamically? To understand your vision here. So I can also send some PRs based upon your vision.

You have a very good idea about DNS events. I would like to see that in the future. But for now, would be nice to see, how would you implement more events dynamically into the Fibratus.

Thank you @rabbitstack 

Yusuf

**Comment by rabbitstack** [comment]
Hi @yusufozturk 

fibratus normalizes event parameters for various purposes, one of them providing a unified and consistent approach to building filter expressions and other constructs. This implies, you would have to change the workflow of how parameter mapping is accomplished, possibly extracting parameter names to external sources such as config files. I think this would have important consequences on fibratus internals and would definitely complicate the code.

I can provide an example of consuming DNS events though, and maybe you could help me with the implementation. Ideally, I would like to preserve current network events and decorate them with DNS payloads. Firstly, you would have to initiate the DNS provider trace, like this:

```go
props := &etw.EventTraceProperties{
		Wnode: etw.WnodeHeader{
			BufferSize: uint32(unsafe.Sizeof(etw.EventTraceProperties{})) + 2*maxStringLen,
			Flags:      etw.WnodeTraceFlagGUID,
		},
		LoggerNameOffset:  uint32(unsafe.Sizeof(etw.EventTraceProperties{})),
		LogFileNameOffset: 0,
		BufferSize:        maxBufferSize,
		LogFileMode:       etw.ProcessTraceModeRealtime,
	}
	handle, err := startTrace(
		"Microsoft-Windows-DNS-Client",
		props,
	)
       guid := syscall.GUID{Data1: 0x1C95126E, Data2: 0x7EEA, Data3: 0x49A9, Data4:[8]byte{0xA3, 0xFE, 0xA3, 0x78, 0xB0, 0x3D, 0xDB, 0x4D }}
	if err := enableTrace(guid, handle, 0); err != nil {
		return fmt.Errorf("couldn't activate DNS client trace: %v", err)
	}
```

The next step is standing up the consumer. This is basically analogous to the current `kstreamc` implementation. You could probably reuse the entire code. In the event callback function, you could get DNS event types like this:

```go
id := evt.Header.EventDescriptor.ID
switch typ{
case 3006, 3009:
	// DNS question
case 3008, 3020, 3018:
	// DNS answer
}
// evt.Header.ProcessID is the pid of the process sending/receiving DNS answers/questions
```
Finally, you can get DNS event parameters with the `produceKparams` method. Example:

```
InterfaceIndex➜ 0, IsAsyncQuery➜ 0, IsNetworkQuery➜ 0, NetworkQueryIndex➜ 0, QueryName➜ stackoverflow.com, QueryOptions➜ 140737489406145, QueryType➜ 28, ServerList➜

AdapterName➜ Wi-Fi, DNSServerAddress➜ 192.168.1.1, InterfaceCount➜ 1, IsParallelNetworkQuery➜ 1, LocalAddress➜ fe80::c064:aca1:f552:7a62;192.168.1.62, NetworkIndex➜ 0, QueryName➜ stackoverflow.com

QueryName➜ stackoverflow.com, QueryOptions➜ 140737489406145, QueryResults➜ , QueryStatus➜ 9501, QueryType➜ 28
```

So, basically, DNS questions map to the `Send` events, and DNS answers are contained within the `Recv` event. We can use the process id as a discriminant to join `Send`/`Recv` events with DNS questions/answers respectively. Then we can simply add a new parameter to `Recv`/`Send` events, something like, `dns`, which would hold the query name, query type, options, etc. Then we would build filters that operate on this parameter:

```
fibratus run net.dns.query matches 'stack.*.com'
```

**Comment by rabbitstack** [comment]
Hi @yusufozturk 

For what's worth, I've put together a PR #107 which adds support for a new collection of events obtained from alternative ETW providers (note that Fibratus mostly relies on the NT Kernel Logger provider). 
It should be quite painless to incorporate events from other ETW providers. Let me know if you need some guidance.

**Comment by yusufozturk** [comment]
@rabbitstack Thanks for the direction.

@namles fyi.
